### PR TITLE
Add save as draft option to assessments

### PR DESCRIPTION
**Pull request by melzreal**
### Description of change

- Add a save and come back to later option to the Assessment flow
- Adds a "submitted" column to each review so we can scope recommendations to submitted and not for show purposes.
- Allows for the recommendation to be saved without enforcing validation as required by ACs

### Story Link
https://trello.com/c/hZvZPEWC/685-when-using-assess-proposal-provide-a-save-as-draft-option

### Design Links
https://beta-bops-prototypes.herokuapp.com/v20/pd-summary
https://beta-bops-prototypes.herokuapp.com/v20/assessment-pd

### CODE REVIEW

File: app/controllers/planning_applications_controller.rb [hunk 1]
```diff
@@ -144,6 +145,17 @@ def view_recommendation
     @recommended_date = @planning_application.recommendations.last.created_at.strftime("%d %b %Y")
   end
 
+  def save_assessment
+    @planning_application.public_comment = params[:planning_application][:public_comment]
+    recommendation = @planning_application.recommendations.build(assessor: current_user)
+    recommendation.assessor_comment = params[:recommendation][:assessor_comment]
+    recommendation.save(validate: false)
```
Comment (benbaumann95) [inline app/controllers/planning_applications_controller.rb:152]: why is this needed ?

**Review by benbaumann95 — COMMENTED** [review]
Looking good so far 👍  I like that you introduced a new state to manage this save as draft functionality. Have pulled the branch locally and all seems to work ok but got a couple questions

File: app/models/planning_application.rb [hunk 2]
```diff
@@ -65,8 +66,16 @@ class PlanningApplication < ApplicationRecord
       transitions from: %i[not_started invalidated in_assessment], to: :in_assessment, guard: :has_validation_date?
     end
 
+    event :save_assessment do
+      transitions from: %i[in_assessment assessment_in_progress], to: :assessment_in_progress
```
[the remark this answers is not in the document]
Comment (melzreal) [reply app/models/planning_application.rb:70]: yes, this happens because we want to allow the save_assessment to be called multiple times whenever someone hits that save draft button - it didn't complain thus far

File: db/migrate/20211231120435_add_time_stamp_to_in_assessment_in_progress.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddTimeStampToInAssessmentInProgress < ActiveRecord::Migration[6.1]
+  def change
+    add_column :planning_applications, :assessment_in_progress_at, :timestamp
```
[the remark this answers is not in the document]
Comment (melzreal) [reply db/migrate/20211231120435_add_time_stamp_to_in_assessment_in_progress.rb:5]: this is for the AASM machine - you can't create an event without the timestamp for the event name. The aasm gem automatically timestamps the event change to a column with the same event name (if you have a look in schema you'll find a bunch of these for the other events)

File: db/migrate/20211231131635_make_assessor_id_optional.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class MakeAssessorIdOptional < ActiveRecord::Migration[6.1]
+  def change
+    change_table :recommendations do |t|
+      t.change_null :assessor_id, true
```
[the remark this answers is not in the document]
Comment (melzreal) [reply db/migrate/20211231131635_make_assessor_id_optional.rb:6]: yes, so my idea was to be more tolerant on this - so we can have a reviewer also save an assessment. I wasn't sure if it made much sense to keep as our flow always implies the review is started by an assessor and then checked by a reviewer - I can take this out, but also didn't see the need to enforce the key dependency if we could be a bit more tolerant 🤷‍♀️

File: app/controllers/planning_applications_controller.rb (hunk 1 above)
Comment (melzreal) [reply app/controllers/planning_applications_controller.rb:152]: A recommendation is made of these fields - the public_comment on the form is associated with the planning application whereas the recommendation has these separate params. I'm using a separate action to save the recommendation because we want validate: false - its an AC on the ticket to not enforce validations when saving as draft

**Review by benbaumann95 — APPROVED** [review]
